cprogs/really.8
.TH really 8 2001-10-21 chiark-backup
.SH NAME
really \- gain privilege or run commands a different user
.SH SYNOPSIS
.B really
.RI [ options ]
.RI [ "command args... " ]
.SH DESCRIPTION
.B really
checks whether the caller is allowed, and if it is it changes its uids
and gids (and perhaps root directory) according to the command line
options and executes the specified command.
.PP
If no options are specified, the uid will be set to 0 and the gids
and root directory will be left unchanged.
.PP
If no command is specified,
.B really
will run
.BR "$SHELL -i" .
.PP
A caller is allowed if it has write access to
.BR /etc/inittab
and is also member of the group
.BR root .
This is most easily achieved by making inittab group-writeable by some
suitable group containing all the appropriate users, and making
.B /etc/inittab
group-owned by that group and group-writeable.  The root group is
perhaps a good choice if it isn't being used for anything else.
.SH OPTIONS
.TP
\fB-u\fR \fIusername\fR | \fB--user\fR \fIusername\fR
Sets the uid, gid, and supplementary group list, according to
.IR username 's
entry in the password and group databases.
.TP
\fB-i\fR \fIusername\fR | \fB--useronly\fR \fIusername\fR
Sets only the uid according to
.IR username 's
entry in the password database.
.TP
\fB-I\fR \fIuid\fR | \fB--uidonly\fR \fIuid\fR
Sets the uid to the numeric value
.I uid
(which need not correspond to any existing user in the password
database).
.TP
\fB-g\fR \fIgroupname\fR | \fB--group\fR \fIgroupname\fR
.I groupname
is looked up in the group database and its gid is appended to the
process's supplementary groups list.  If this is the first gid
specified it will also be set as the primary gid.
.TP
\fB-G\fR \fIgid\fR | \fB--gid\fR \fIgid\fR
.I gid
is appended to the process's supplementary groups list.
.RI ( gid
need not correspond to any existing group in the group database.)  If
this is the first gid specified it will also be set as the primary
gid.
.TP
\fB-z\fR | \fB--groupsclear\fR
Clears the process's supplementary groups list.  When using this
option you must also specify
.B -g
or
.BR -G .
The process's groups will then be exactly those specified.  The
relative position of
.B -z
in the argument list is not relevant.
.TP
\fB-R\fR \fIroot-dir\fR | \fB--chroot\fR \fIroot-dir\fR
The program will have its root directory set to
.IR root-dir .

.BR "Do not use this option unless you know what you are doing" :
Unlike chroot(8), the current working directory will remain unchanged.
This means that if the current directory isn't underneath the
specified new root, the program will still be able to access files
outside the new root by using relative pathnames.  If this isn't
what you want, please use the chroot utility instead.
.TP
.B \-\-
Indicates the end of the options.  The next argument (if present) will
be interpreted as the command name, even if it starts with a hyphen.
.SH SECURITY CONSIDERATIONS
.B really
is designed so that installing it setuid root is extremely unlikely to
compromise the security of any system.  It will check using
.BR access (2)
whether the real user is allowed to write to
.B /etc/inittab
and if this check fails
.B really
will exit without even attempting to parse its command line.
.PP
.B really
is
.B not
designed to be resistant to malicious command line arguments.  Do not
allow untrusted processes to pass options to really, or to specify the
command to be run.  Whether it is safe to allow relatively untrusted
processes to pass options to the command which is to be run depends on
the behaviour of that command and its security status.
.PP
Attempting to use
.B really
to drop privilege is dangerous unless the calling environment is very
well understood.  There are many inherited process properties and
resources which might be used by the callee to escalate its privilege
to that of the (root-equivalent) caller.  For this function, it is
usually better to use
.B userv
if possible.
.SH ENVIRONMENT
.B really
does not manipulate the environment at all.  The calling program is
run in exactly the same environment as the caller passes to
.BR really .
In particular,
.B really
will not add
.B sbin
directories to
.B PATH
so
.BR really -enabled
accounts will usually need to have these directories on their
configured
.B PATH
to start with.
.PP
.B SHELL
is used to find the default shell to use in interactive mode (ie, when
no command is specified).
.SH AUTHOR
This version of
.B really
was written by Ian Jackson <ian@chiark.greenend.org.uk>.
.PP
It and this manpage are Copyright (C) 1992-5,2004,2013 Ian Jackson
<ian@chiark.greenend.org.uk>.
.PP
.B really
is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 3,
or (at your option) any later version.
.PP
.B really
is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public
License along with this file; if not, consult the Free Software
Foundation's website at www.fsf.org, or the GNU Project website at
www.gnu.org.
.SH AVAILABILITY
.B really
is currently part of
.B chiark-utils
and is available for download from
ftp.chiark.greenend.org.uk in /users/ian/chiark-utils/,
in source and pre-compiled binary form, and also from Ian Jackson's
cvsweb.
.SH "SEE ALSO"
.BR userv (1),
.BR access (2),
.BR setresuid (2),
.BR setresgid (2),
.BR setgroups (2)


debian/changelog
chiark-utils (8.0.0-0deepin1) unstable; urgency=medium

  * feat: add sw64 support 

 -- hudeng <hudeng@deepin.org>  Tue, 21 Oct 2025 17:32:41 +0800

chiark-utils (8.0.0) unstable; urgency=medium

  * Rewrite the packaging to use `debhelper` and `dh` in as standard a way
    as possible.  As a consequence, support root-less builds.
    No functional changes to the installed packages.
    Closes: #1089303.  [Niels Thykier]

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Sun, 12 Jan 2025 20:19:58 +0000

chiark-utils (7.0.3) unstable; urgency=medium

  * Team upload.
  * git-cache-proxy: Handle extra-parameters in git protocol (Closes: #1040476).
    Thanks to Anthony PERARD for the patch.
  * git-branchmove: Always pass --no-verify to 'git push' (Closes: #1080082).

 -- Sean Whitton <spwhitton@spwhitton.name>  Thu, 26 Sep 2024 17:07:41 +0100

chiark-utils (7.0.2) unstable; urgency=medium

  * debian/rules: Pass -D_TIME_BITS=64 too.  Closes: #1078190.
  * Change file size type to unsigned long long.  Closes: #1075821.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Thu, 15 Aug 2024 22:28:46 +0100

chiark-utils (7.0.1) unstable; urgency=medium

  * backup/snaprsync: add -AX to the rsync rune
  * Enable LFS and 64-bit time_t support.  Closes: #1067157.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Thu, 04 Jul 2024 12:53:58 +0100

chiark-utils (7.0.0) unstable; urgency=medium

  New facility:
  * prefork-interp, Proc::Prefork::Interp: new facility

  Improvements:
  * expire-iso8601: Replace with new implementation.  Closes:#862897.

  Bugfixes:
  * acctdump: Fix manual page section.
  * git-cache-proxy: fix for new git.  Closes: #1006695.  [Anthony Perard]
  * git-branchmove: install manpage..  Closes: #946364.  [Sean Whitton]
  * hexterm: Update #! and Suggest tcl8.6.  Closes: #905474.

  Build and packaging improvements:
  * Fix cross build.  Closes: #951242.  [Helmut Grohne]
  * copyright: Be clear that SPDX indications are licence grants
  * Add lintian override for python dependency
  * Add lintian override for chiark-really
  * fix priority "extra" to "optional"
  * Fix build-dependency on debhelper.
  * Set source format to 3.0 (native)
  * Add Vcs-* fields.
  * Update Standards-Version
  * timespeccmp.h: Move use to just cgi-fcgi-interp

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 22 Aug 2022 00:06:57 +0100

chiark-utils (6.1.2+nmu1) unstable; urgency=medium

  * Non-maintainer upload.
    - Maintainer has approved this via personal communication.
  * Bump debhelper compat to 13 (Closes: #965453).

 -- Sean Whitton <spwhitton@spwhitton.name>  Tue, 21 Jul 2020 08:17:49 -0700

chiark-utils (6.1.2) unstable; urgency=medium

  * git-cache-proxy: Recheck gc timestamp after taking lock
  * git-cache-proxy: Tolerate failure to write gc stamp
  * git-cache-proxy: Do not update timestamp due to housekeeping gc

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Wed, 12 Feb 2020 16:10:02 +0000

chiark-utils (6.1.2~citrix2~) unstable; urgency=medium

  * git-cache-proxy: Move $SIG{ALRM} setting to global
  * git-cache-proxy: Abandon fetch attempt if client disconnects
  * git-cache-proxy: Use open-coded fork/exec for git gc

 -- Ian Jackson <ian.jackson@citrix.com>  Tue, 11 Feb 2020 18:34:58 +0000

chiark-utils (6.1.1) unstable; urgency=medium

  * fishdescriptor: Use Python "errno" module
  * git-cache-proxy: Periodically run `git gc --quiet'
  * git-cache-proxy: Document some options in the head comment

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Tue, 11 Feb 2020 16:37:31 +0000

chiark-utils (6.0.5~citrix1) unstable; urgency=medium

  * fishdescriptor: Use Python "errno" module, not "os.errno"
    (which is not in modern python3 apparently).

 -- Ian Jackson <ian.jackson@citrix.com>  Fri, 20 Dec 2019 17:50:02 +0000

chiark-utils (6.1.0~iwj1) unstable; urgency=medium

  * really: Writeability of rc.local suffices, too.
    (Non-sysvinit compatibility.)

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 12 Aug 2019 10:59:02 +0100

chiark-utils (6.1.0) unstable; urgency=medium

  * Non-maintainer upload.
    - Maintainer has approved this via personal communication.
  * git-branchmove: rewrite in perl (Closes: #914398, #914399)
    - Add dependencies on libgit-wrapper-perl, libtry-tiny-perl to
      bin:chiark-scripts.
  * git-branchmove: new --detach feature.

 -- Sean Whitton <spwhitton@spwhitton.name>  Sat, 07 Dec 2019 14:10:26 -0700

chiark-utils (6.0.4) unstable; urgency=medium

  * sync-accounts: Fix perl syntax error.  Closes:#865985.
  * changelog: Document bug number for bugfix in 6.0.4~citrix1.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Fri, 12 Apr 2019 11:48:00 +0100

chiark-utils (6.0.4~citrix1) unstable; urgency=medium

  * fishdescriptor: cast __errno_location correctly.  Closes:#926858.

 -- Ian Jackson <ian.jackson@citrix.com>  Mon, 08 Apr 2019 17:03:47 +0100

chiark-utils (6.0.3) unstable; urgency=medium

  * Upload to Debian unstable.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Tue, 20 Nov 2018 18:25:53 +0000

chiark-utils (6.0.3~citrix1) unstable; urgency=medium

  * git-branchmove: Avoid transporting tags about
  * fishdescriptor: Cope if donor is in a chroot without /proc
  * fishdescriptor: Cope if only donor has /run/user/UID

 -- Ian Jackson <ian.jackson@citrix.com>  Mon, 19 Nov 2018 16:11:55 +0000

chiark-utils (6.0.2) unstable; urgency=medium

  * Rebuild with arch:all packages included in the upload.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 11 Jun 2018 14:35:08 +0100

chiark-utils (6.0.1) unstable; urgency=low

  fishdescriptor:
  * Fix numerous bugs, affecting the `exec' feature with nontrivial fd
    specifications.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 11 Jun 2018 14:31:24 +0100

chiark-utils (6.0.0) unstable; urgency=medium

  xbatmon-simple:
  * Minor improvements to the .txt doc.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Fri, 20 Apr 2018 16:56:04 +0100


cprogs/really.c
/*
 * really.c - program for gaining privilege
 *
 * Copyright (C) 1992-3 Ian Jackson <ian@davenant.greenend.org.uk>
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3,
 * or (at your option) any later version.
 *
 * This is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this file; if not, consult the Free Software
 * Foundation's website at www.fsf.org, or the GNU Project website at
 * www.gnu.org.
 */

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <unistd.h>
#include <pwd.h>
#include <grp.h>
#include <sys/types.h>
#include <errno.h>

#include "myopt.h"

void usagemessage(void) {
  if (fputs("usage: really [<really-option> ...] [--]"
            " [<command> [<argument/option> ...]]\n"
            "really-options specifying the user:\n"
            " if no options given, set the uid to 0;\n"
            " -u|--user <username>     also sets their default group list\n"
            " -i|--useronly <username> } set the uid\n"
            " -I|--uidonly <uid>       }  but inherits the group list\n"
            "really-options specifying the group:\n"
            " -z|--groupsclear         only groups specified are to be used\n"
            " -g|--group <groupname>   } add this to\n"
            " -G|--gid <gid>           }  the group list\n"
            "other really-options:\n"
	    " -h|--help                display this message\n"
	    " -R|--chroot <dir>        chroot (but *not* chdir - danger!)\n",
            stderr) == EOF) { perror("write usage"); exit(-1); }
}

static const char *opt_user, *opt_useronly, *opt_chroot;
static int opt_groupsclear= 0, opt_ngids= 0, opt_uidonly= -1;
static int opt_gids[512];

static void af_uidonly(const struct cmdinfo *cip, const char *value) {
  unsigned long ul;
  char *ep;

  ul= strtoul(value,&ep,10);
  if (*ep) { fprintf(stderr,"bad uid `%s'\n",value); exit(-1); }
  opt_uidonly= ul;
}

static void af_group(const struct cmdinfo *cip, const char *value) {
  struct group *gr;

  if (opt_ngids >= sizeof(opt_gids)/sizeof(opt_gids[0]))
    badusage("too many groups specified");
  gr= getgrnam(value);
  if (!gr) { fprintf(stderr,"unknown group `%s'\n",value); exit(-1); }
  opt_gids[opt_ngids++]= gr->gr_gid;
}

static void af_gid(const struct cmdinfo *cip, const char *value) {
  char *ep;
  unsigned long ul;

  if (opt_ngids >= sizeof(opt_gids)/sizeof(opt_gids[0]))
    badusage("too many gids specified");
  ul= strtoul(value,&ep,0);
  if ((*ep) || (uid_t)ul != ul || ul>INT_MAX) badusage("bad gid `%s'",value);
  opt_gids[opt_ngids++]= ul;
}

static void af_help(const struct cmdinfo *cip, const char *value) {
  usagemessage(); exit(0);
}

static const struct cmdinfo cmdinfos[]= {
  { "user",         'u',  1,  0, &opt_user,          0,           },
  { "useronly",     'i',  1,  0, &opt_useronly,      0            },
  { "uidonly",      'I',  1,  0, 0,                  af_uidonly   },
  { "groupsclear",  'z',  0,  &opt_groupsclear, 0,   0,        1  },
  { "group",        'g',  1,  0, 0,                  af_group     },
  { "gid",          'G',  1,  0, 0,                  af_gid       },
  { "chroot",       'R',  1,  0, &opt_chroot,        0            },
  { "help",         'h',  0,  0, 0,                  af_help      },
  {  0,              0                                            }
};

#ifdef REALLY_CHECK_FILE
static int checkroot(void) {
  int r;
  r= access(REALLY_CHECK_FILE,   W_OK);
  if (!r) return 0;
#ifdef REALLY_CHECK_FILE_2
  r= access(REALLY_CHECK_FILE_2, W_OK);
  if (!r) return 0;
  /* If all fails we return the errno from file _2 */
#endif /*REALLY_CHECK_FILE_2*/
  return -1;
}
#endif
#ifdef REALLY_CHECK_GID
static int checkroot(void) {
  gid_t groups[512];
  int r, i;

  r= getgid(); if (r==REALLY_CHECK_GID) return 0;
  if (r<0) { perror("getgid check"); exit(-1); }
  r= getgroups(sizeof(groups)/sizeof(groups[0]),groups);
  if (r<0) { perror("getgroups check"); exit(-1); }
  for (i=0; i<r; i++)
    if (groups[i] == REALLY_CHECK_GID) return 0;
  return -1;
}
#endif
#ifdef REALLY_CHECK_NONE
static int checkroot(void) {
  return 0;
}
#endif

int main(int argc, const char *const *argv) {
  struct passwd *pw= 0;
  gid_t groups[512];
  int i, j, ngroups, ngroups_in, maingid, orgmaingid, mainuid, orgmainuid, r;
  const char *cp;
  
  orgmainuid= getuid();
  if (orgmainuid && checkroot()) { perror("sorry"); exit(-1); }
  myopt(&argv,cmdinfos);

  if (opt_groupsclear && !opt_ngids)
    badusage("-z|--groupsclear must be accompanied by some groups");
  if (opt_user && (opt_useronly || opt_uidonly!=-1))
    badusage("-u|--user may not be used with -i|--useronly or -I|--uidonly");
  if (opt_user && opt_groupsclear)
    badusage("-u|--user may not be used with -z|--groupsclear");
  if (opt_uidonly != -1 && (uid_t)opt_uidonly != opt_uidonly)
    badusage("-I|--uidonly value %d is out of range for a uid",opt_uidonly);

  if (!opt_user && !opt_useronly && opt_uidonly==-1 && !opt_ngids) {
    opt_uidonly= 0;
  }
  if (opt_user || opt_useronly) {
    cp= opt_user ? opt_user : opt_useronly;
    pw= getpwnam(cp);
    if (!pw) { fprintf(stderr,"unknown user `%s'\n",cp); exit(-1); }
    opt_uidonly= pw->pw_uid;
  }
  if (opt_chroot) {
    if (chroot(opt_chroot)) { perror("chroot failed"); exit(-1); }
  }
  orgmaingid= getgid();
  if (orgmaingid<0) { perror("getgid failed"); exit(-1); }
  if (opt_user) {
    r= initgroups(opt_user,pw->pw_gid);
    if (r) { perror("initgroups failed"); exit(-1); }
    maingid= pw->pw_gid;
  } else {
    maingid= -1;
  }
  if (opt_groupsclear) {
    ngroups= 0;
    if (opt_ngids > sizeof(groups)/sizeof(groups[0])) {
      fputs("too many groups to set\n",stderr);
      exit(-1);
    }
  } else {
    ngroups= getgroups(0,0);
    if (ngroups<0) { perror("getgroups(0,0) failed"); exit(-1); }
    if (ngroups+opt_ngids > sizeof(groups)/sizeof(groups[0])) {
      fputs("too many groups already set for total to fit\n",stderr);
      exit(-1);
    }
    ngroups= getgroups(ngroups,groups);
    if (ngroups<0) { perror("getgroups failed"); exit(-1); }
  }
  if (opt_ngids) {
    maingid= opt_gids[0];
  }
  if (opt_ngids || opt_groupsclear) {
    ngroups_in= ngroups; ngroups= 0;
    for (i=0; i<ngroups_in; i++) {
      for (j=0; j<ngroups && groups[j] != groups[i]; j++);
      if (j<ngroups) continue;
      groups[ngroups++]= groups[i];
    }
    for (i=0; i<opt_ngids; i++) {
      for (j=0; j<ngroups && groups[j] != opt_gids[i]; j++);
      if (j<ngroups) continue;
      groups[ngroups++]= opt_gids[i];
    }
    r= setgroups(ngroups,groups);
    if (r) { perror("setgroups failed"); exit(-1); }
  }
  if (maingid != -1) {
    r= setgid(maingid); if (r) { perror("setgid failed"); exit(-1); }
    r= setgid(maingid); if (r) { perror("2nd setgid failed"); exit(-1); }
  }
  if (opt_uidonly != -1) {
    mainuid= opt_uidonly;
  } else {
    mainuid= orgmainuid;
  }
  r= setuid(mainuid); if (r) { perror("setuid failed"); exit(-1); }
  r= setuid(mainuid); if (r) { perror("2nd setuid failed"); exit(-1); }
  if (mainuid != 0) {
    r= seteuid(0); if (r>=0) { fputs("could seteuid 0",stderr); exit(-1); }
    if (errno != EPERM) {
      perror("unexpected failure mode for seteuid 0"); exit(-1);
    }
  }
  r= getuid(); if (r<0) { perror("getuid failed"); exit(-1); }
  if (r != mainuid) { fputs("getuid mismatch",stderr); exit(-1); }
  r= geteuid(); if (r<0) { perror("geteuid failed"); exit(-1); }
  if (r != mainuid) { fputs("geteuid mismatch",stderr); exit(-1); }
  if (maingid != -1) {
    for (i=0; i<ngroups && maingid != groups[i]; i++);
    if (i>=ngroups && maingid != orgmaingid) {
      r= setgid(orgmaingid);
      if (r>=0) { fputs("could setgid back",stderr); exit(-1); }
      if (errno != EPERM) {
        perror("unexpected failure mode for setgid back"); exit(-1);
      }
    }
    r= getgid(); if (r<0) { perror("getgid failed"); exit(-1); }
    if (r != maingid) { fputs("getgid mismatch",stderr); exit(-1); }
    r= getegid(); if (r<0) { perror("getegid failed"); exit(-1); }
    if (r != maingid) { fputs("getegid mismatch",stderr); exit(-1); }
  }
  if (!*argv) {
    cp= getenv("SHELL");
    if (!cp) cp= "sh";
    execlp(cp,cp,"-i",(const char*)0);
  } else {
    execvp(argv[0],(char**)argv);
  }
  perror("exec failed");
  exit(-1);
}
